File: cpp/Mod_nonRecursiveCheck.cpp
#pragma once

#include "Mod_sizeLogspace.cpp"
#include "Mod_firstChild.cpp"

int numChildren(Tree t1, Node n1){
	int cnt = 0;
	if(isLeaf(t1, n1))
	   return 0;
	
	Node n = firstChild(t1, n1);
	cnt++;
	
	while(hasNextSibling(t1, n)){
		n = nextSibling(t1, n);
		cnt++;
	}
	return true;
}

int countNumber(Tree t, Node root, int number){
	int cnt = 0;
	Node n = firstChild(t, root);
	int size = sizeLogspace(t, n);
	if(size == number){
		cnt++;
	}
	
	while(hasNextSibling(t, n)){
		n = nextSibling(t, n);
		size = sizeLogspace(t, n);
		if(size == number){
			cnt++;
		}
	}
	
	return cnt;
}

bool lexikoLess(Tree t1, Tree t2, Node n1, Node n2, int numChildren){
	int act = 0;
	while(numChildren != 0){
		int c1 = countNumber(t1, n1, act);
		int c2 = countNumber(t2, n2, act);
		if(c1>c2)
			return true;
		else if(c1<c2)
			return false;
		else{
			numChildren -= c1;
			act++;
		}
	}
	return false;
}

bool nonRecursiveCheck(Tree t1, Tree t2, Node n1, Node n2){
	int s1 = sizeLogspace(t1, n1);
	int s2 = sizeLogspace(t2, n2);
	int num1 = numChildren(t1, n1);
	int num2 = numChildren(t2, n2);
	if( s1 < s2)
		return true;
	if(s1 == s2 && num1 < num2)
		return true;
	if(s1 == s2 && num1 == num2 && lexikoLess(t1, t2, n1, n2, num1))
		return true;
	
	return false;
}	  
	  

File: hs/Mod_nonRecursiveCheck.hs
module Mod_nonRecursiveCheck ( nonRecursiveCheck ) where
import Prelude hiding (and, or) 
import Debug.Trace
import Splib
import Types

import Mod_hasNextSibling
import Mod_nextSibling
import Mod_firstChild
import Mod_sizeLogspace

numChildrenInit :: Tree -> Node -> Int
numChildrenInit t n
 | isLeaf t n = 0
 | otherwise = numChildren t (firstChild t n) 1

numChildren :: Tree -> Node -> Int -> Int
numChildren t n cnt
 | hasNextSibling t n = numChildren t (nextSibling t n) (cnt+1)
 | otherwise = cnt
 
countNumberInit :: Tree -> Node -> Int -> Int
countNumberInit t n num
 | sizeLogspace t firstC == num = countNumber t firstC num 1
 | otherwise = countNumber t firstC num 0
    where
     firstC = firstChild t n
 
countNumber :: Tree -> Node -> Int -> Int -> Int
countNumber t n num cnt
 | hasNextSibling t n && sizeLogspace t next == num = countNumber t next num (cnt+1)
 | hasNextSibling t n = countNumber t next num cnt
 | otherwise = cnt
  where
   next = nextSibling t n
 
lexikoLess :: Tree -> Tree -> Node -> Node -> Int -> Int -> Bool
lexikoLess t1 t2 n1 n2 numChilds act
 | numChilds == 0 = False
 | c1>c2 = True
 | c1<c2 = False
 | otherwise = lexikoLess t1 t2 n1 n2 (numChilds-c1) (act+1)
   where
    c1 = countNumberInit t1 n1 act
    c2 = countNumberInit t2 n2 act

nonRecursiveCheck :: Tree -> Tree -> Node -> Node -> Bool
nonRecursiveCheck t1 t2 n1 n2
 | s1<s2 = True
 | s1 == s2 && num1<num2 = True
 | s1 == s2 && num1== num2 && lexikoLess t1 t2 n1 n2 num1 0 = True
 | otherwise = False
   where
    s1 = sizeLogspace t1 n1
    s2 = sizeLogspace t2 n2
    num1 = numChildrenInit t1 n1
    num2 = numChildrenInit t2 n2